packages/@aws-cdk/aws-pipes-alpha/lib/source.ts

```diff
@@ -99,4 +99,9 @@ export interface ISource {
    * Grant the pipe role read access to the source.
    */
   grantRead(grantee: IRole): void;
+
+  /**
+   * Grant the pipe role write access to the dead-letter target.
+   */
+  grantDlqPush?(grantee: IRole): void;
```

Comment:
@nmussy Making this optional solved a whole lot of problems.

Reply:
I don't know if I'm a fan of having this be an optional method, because you want sources that will have a DLQ property to implement `grantDlqPush`, and those that don't not to have that method. You're not ensuring either with that pattern. I was thinking of doing something like this instead:

```ts
interface ISource {
  /**
   * The ARN of the source resource.
   */
  readonly sourceArn: string;

  /**
   * Bind the source to a pipe.
   */
  bind(pipe: IPipe): SourceConfig;

  /**
   * Grant the pipe role read access to the source.
   */
  grantRead(grantee: IRole): void;
}

interface ISourceWithDlq extends ISource {
  grantDlqPush?(grantee: IRole): void;
}

abstract class SourceWithDlq implements ISourceWithDlq {
  readonly sourceArn: string;
  private readonly deadLetterTargetArn: string;

  abstract grantDlqPush?(grantee: IRole): void;
  abstract bind(pipe: IPipe): SourceConfig;
  abstract grantRead(grantee: IRole): void;
}

class DynamoDBSource extends SourceWithDlq {}
```

This way, you do not need to check for whether or not the method is implemented directly there:
https://github.com/msambol/aws-cdk/blob/8f5df5c453fe60409f2636d90362a22033f3e060/packages/%40aws-cdk/aws-pipes-alpha/lib/pipe.ts#L236-L238

You can instead make a guard to ensure that you have a `SourceWithDlq`:

```ts
const isSourceWithDlq = (source: ISource): source is SourceWithDlq => !!(source as SourceWithDlq).grantDlqPush;
```